### ISSUE TITLE
[consistent-type-assertions] `allow-as-parameter` doesn't work in JSX

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#2587]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```sh
npm install @types/react
```
```JSON
{
  "rules": {
		"@typescript-eslint/consistent-type-assertions": [
			"error",
			{
				"assertionStyle": "as",
				"objectLiteralTypeAssertions": "allow-as-parameter"
			}
		]
  }
}
```

```tsx
<span
	style={{
		'--my-prop': 'hello'
	} as React.CSSProperties}
/>
```

Code explanation: https://stackoverflow.com/a/54128069/288906

**Expected Result**

No issue because it's not an assignment and therefore I can't use `const x: T = { ... }`

Also I think this case should be covered by the `allow-as-parameter` since it's transpiled to

```js
React.createElement("span", {
  style: {
    '--my-prop': 'hello'
  }
});
```

**Actual Result**
```
  ✖  50:13  Always prefer const x: T = { ... }.  @typescript-eslint/consistent-type-assertions
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `14.4.0` |
